ballista/rust/executor/src/executor_server.rs

```diff
@@ -0,0 +1,290 @@
+}
+
+async fn register_executor(
+    scheduler: &mut SchedulerGrpcClient<Channel>,
+    executor_meta: &ExecutorRegistration,
+    specification: &ExecutorSpecification,
+) -> Result<(), BallistaError> {
+    let result = scheduler
+        .register_executor(RegisterExecutorParams {
+            metadata: Some(executor_meta.clone()),
+            specification: Some(specification.clone().into()),
+        })
+        .await?;
+    if result.into_inner().success {
+        Ok(())
+    } else {
+        Err(BallistaError::General(
+            "Executor registration failed!!!".to_owned(),
+        ))
+    }
+}
+
+#[derive(Clone)]
+pub struct ExecutorServer {
+    _start_time: u128,
+    executor: Arc<Executor>,
+    executor_meta: ExecutorRegistration,
+    scheduler: SchedulerGrpcClient<Channel>,
+    executor_env: ExecutorEnv,
+}
+
+#[derive(Clone)]
+struct ExecutorEnv {
+    tx_task: mpsc::Sender<TaskDefinition>,
+}
+
+unsafe impl Sync for ExecutorEnv {}
+
+impl ExecutorServer {
+    fn new(
+        scheduler: SchedulerGrpcClient<Channel>,
+        executor: Arc<Executor>,
+        executor_meta: ExecutorRegistration,
+        executor_env: ExecutorEnv,
+    ) -> Self {
+        Self {
+            _start_time: SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap()
+                .as_millis(),
+            executor,
+            executor_meta,
+            scheduler,
+            executor_env,
+        }
+    }
+
+    async fn heartbeat(&self) {
+        // TODO Error handling
+        self.scheduler
+            .clone()
+            .send_heart_beat(SendHeartBeatParams {
+                metadata: Some(self.executor_meta.clone()),
+                state: Some(self.get_executor_state().await.into()),
+            })
+            .await
+            .unwrap();
+    }
+
+    async fn run_task(&self, task: TaskDefinition) -> Result<(), BallistaError> {
+        let task_id = task.task_id.unwrap();
+        let task_id_log = format!(
+            "{}/{}/{}",
+            task_id.job_id, task_id.stage_id, task_id.partition_id
+        );
+        info!("Start to run task {}", task_id_log);
+
+        let plan: Arc<dyn ExecutionPlan> = (&task.plan.unwrap()).try_into().unwrap();
+        let shuffle_output_partitioning =
+            parse_protobuf_hash_partitioning(task.output_partitioning.as_ref())?;
+
+        let execution_result = self
+            .executor
+            .execute_shuffle_write(
+                task_id.job_id.clone(),
+                task_id.stage_id as usize,
+                task_id.partition_id as usize,
+                plan,
+                shuffle_output_partitioning,
+            )
+            .await;
+        info!("Done with task {}", task_id_log);
+        debug!("Statistics: {:?}", execution_result);
+
+        // TODO use another channel to update the status of a task set
+        self.scheduler
+            .clone()
+            .update_task_status(UpdateTaskStatusParams {
```

Comment:
I probably missed this in the code, what happens when the executor crashed and never report an update for its assigned task? Not a blocker to this PR, just curious how it's being handled.

Reply:
Actually how to deal with the executor crashed case is in the TODO list. Like Spark, if the some task fails, a speculative task will be scheduled by the scheduler. 

This initial version does put much attention to the error handling. However, with this framework, it's doable.